File: tax_progressive/tax.py
def calculate_tax(income, tax_brackets):
    tax_result = 0
    remaining_income = income
    calculation_list = []

    for bracket in tax_brackets:
        bracket_limit, tax_rate = bracket
        if remaining_income > 0:
            taxable_amount = min(remaining_income, bracket_limit)
            tax_result += taxable_amount * tax_rate
            remaining_income -= taxable_amount

            # printable format
            formatted_taxable = format(int(taxable_amount), ",").replace(",", ".")
            formatted_tax_rate = "{:.0%}".format(tax_rate)
            calculation_list.append(f"{formatted_taxable} * {formatted_tax_rate}")
        else:
            break

    return calculation_list, tax_result

tax_brackets_list = [(30_000_000, 0.05), (100_000_000, 0.10), (float('inf'), 0.20)]

# Input the income
income_input = float(input("input: "))
printable_calculation, tax = calculate_tax(income_input, tax_brackets_list)
formatted_tax = format(int(tax), ",").replace(",", ".")
print("output: ", " + ".join(printable_calculation), "=", formatted_tax)

File: tax_progressive/tax.js
function calculateTax(income, taxBrackets) {
    let taxResult = 0;
    let remainingIncome = income;
    const calculationList = [];

    for (const bracket of taxBrackets) {
        const [bracketLimit, taxRate] = bracket;
        if (remainingIncome > 0) {
            const taxableAmount = Math.min(remainingIncome, bracketLimit);
            taxResult += taxableAmount * taxRate;
            remainingIncome -= taxableAmount;

            // Printable format
            const formattedTaxable = taxableAmount.toLocaleString('en-US', { useGrouping: false });
            const formattedTaxRate = (taxRate * 100).toFixed(0) + '%';
            calculationList.push(`${formattedTaxable.replace(',', '.') + ' * ' + formattedTaxRate}`);
        } else {
            break;
        }
    }

    return [calculationList, taxResult];
}

const taxBracketsList = [[30000000, 0.05], [100000000, 0.10], [Infinity, 0.20]];

const readline = require('readline');
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question("input: ", (incomeInput) => {
    const income = parseFloat(incomeInput);
    const [printableCalculation, tax] = calculateTax(income, taxBracketsList);
    const formattedTax = tax.toLocaleString('en-US', { useGrouping: false });

    console.log("output: " + printableCalculation.join(' + ') + ' = ' + formattedTax.replace(',', '.'));
    rl.close();
});
